src/models/base_model.py

```diff
@@ -0,0 +1,234 @@
+
+        Returns:
+            output: Dict of module hook name to its output tensor.
+        """
+        output = self._feature_outputs[device]
+        # clear after reading
+        self._feature_outputs[device] = OrderedDict()  
+        return output
+
+
+class BaseModel(nn.Module, ABC):
+    """Base model for all TorchOk Models - Backbone, Neck, Pooling and Head.
+
+    This class supports adding feature hooks to some model layers.
+    Class has feature info to describe hooks.
+    To create hooks, method `self.get_features_info()` must be overriden, and the function `self.create_hooks()` 
+    must be called in the end line of initialization method of inherited class!
+    Method `self.create_hooks(*args, **kwargs)` called `self.get_features_info(*args, **kwargs)` with 
+    the same parameters! 
+    Also class contains method get_output_channels which returns forward pass and hooks output channels.
+
+    Example:
+        >>> class ModelWithHooks(BaseModel):
+        >>>     def __init__(self, input_channel: int = 3):
+        >>>         super().__init__()
+        >>>         self.out_features = [5, 10, 15]
+        >>>         self.conv1 = nn.Conv2d(input_channel, 5, 3)
+        >>>         self.conv2 = nn.Conv2d(5, 10, 3)
+        >>>         self.conv3 = nn.Conv2d(10, 15, 3)
+        >>>         # Call create_hooks after all modules are initialized, with parameters as in get_features_info.
+        >>>         self.create_hooks(output_channels=[10, 15], module_names=['conv2', 'conv3'])
+        >>>
+        >>>     def get_features_info(self, output_channels, module_names):
+        >>>         # Overwrite get_features_info.
+        >>>         features_info = []
+        >>>         for num_channels, module_name in zip(output_channels, module_names):
+        >>>             feature_info = FeatureInfo(module_name=module_name, num_channels=num_channels, stride=1)
+        >>>             features_info.append(feature_info)
+        >>>         return features_info
+        >>>
+        >>>     def get_forward_output_channels(self):
+        >>>         return self.out_features[-1]
+        >>>
+        >>> model = ModelWithHooks()
+        >>> print(model.get_output_channels())
+        (15, [10, 15])
+    """
+    def __init__(self):
+        """Initialize BaseModel.
+        
+        Set features_info and feature_hooks as None.
+        """
+        super().__init__()
+        self.__features_info = None
+        self.__feature_hooks = None
+
+    def get_features_info(self, *args: Any, **kwargs: Any) -> List[FeatureInfo]:
+        """Initialize feature hooks.
+        
+        It should be overriden in an inherited class if a user is expected to have access to feature hooks 
+        of the implemented Module.
+
+        Retruns: Hooks FeatureInfo list.
+        """
+        return []
+
+    def create_hooks(self, *args: Any, **kwargs: Any) -> Tuple[List[FeatureInfo], FeatureHooks]:
```

Comment:
Why does a user need to pass arguments here? I haven't seen any case requiring this

Reply:
The reason is user's reluctance to put hook parameters into class fields